File: backend/src/domain/entities/algorithm.py
# Representa um algoritmo disponível no sistema.
# Essa classe é puro dado — sem lógica, sem framework.
# Qualquer parte do sistema que precisar falar sobre um algoritmo usa isso aqui.

from dataclasses import dataclass


@dataclass(frozen=True)
class Algorithm:
    id: str           # ex: "bubble_sort"
    name: str         # ex: "Bubble Sort"
    category: str     # ex: "array", "graph", "tree"
    description: str
    time_complexity: str   # ex: "O(n²)"
    space_complexity: str  # ex: "O(1)"

File: frontend/src/domain/Algorithm.ts
// Espelha as entidades do backend em TypeScript.
// Qualquer parte do frontend que precisar falar sobre um algoritmo usa esse tipo.

export interface Algorithm {
  id: string
  name: string
  category: string
  description: string
  time_complexity: string
  space_complexity: string
}

export interface ExecutionStep {
  step_number: number
  description: string
  state: Record<string, unknown>   // varia por algoritmo
  highlighted: (number | string)[] // índices ou nomes de nós
  is_final: boolean
}

export interface ExecutionResult {
  algorithm_id: string
  total_steps: number
  steps: ExecutionStep[]
}
